features/56-login-with-custom-token/api.md

```diff
@@ -0,0 +1,160 @@
+    }
+
+    @Override
+    public void onAuthFail(Error error) {
+        Log.e("Skygear loginWithCustomToken", "onAuthFail: Reason: " + error.getMessage());
+    }
+});
+```
+
+### Sample Codes (User's server)
+
+1. Get the token secret from portal, and setup environment variable in your server `SKYGEAR_CUSTOM_TOKEN_SECRET`
+
+2. Generate create token
+
+**Python (With SDK)**
+
+```python
+from skygear.utils.auth import create_custom_token
+
+custom_token = create_custom_token("your server user id", profile)
+
+```
+
+**Rudy (Without SDK)**
+
+- Install [ruby-jwt](https://github.com/jwt/ruby-jwt)
+
+```ruby
+custom_token_secret = "Custom token secret from portal"
+user_id = "unique identifier of user in your server, string between 1-36 characters long"
+now_seconds = Time.now.to_i
+
+payload = {
+    :sub => user_id,
+    :iat => now_seconds,
+    :exp => now_seconds+(60*60), # expire within 1 hour
+    :skygear_profile => {}
+}
+
+JWT.encode payload, custom_token_secret, "HS256"
+```
+
+### List of APIs (Client)
+
+**py-skygear**
+
+```
+skygear.utils.auth.create_custom_token(principle_id, profile={})
+```
+
+**JS**
+
+```js
+skygear.auth.loginWithCustomToken(customToken: String): Promise<Record>
+```
+
+**iOS (swift)**
+
+```obj-c
+- (void)updateWithCustomToken:(NSString *)customToken
+            completionHandler:(SKYContainerUserOperationActionCompletion _Nullable)completionHandler;
+```
+
+**Android (java)**
+
+```java
+public void loginWithCustomToken(java.lang.String customToken,
+                                 AuthResponseHandler handler);
+```
+
+### List of APIs (User's server)
+
+**py-skygear**
+
+```
+skygear.utils.auth.create_custom_token(principle_id, profile={})
+```
+
+**JS SDK**
+
+```js
+import { createCustomToken } from 'skygear-build/packages/skygear-core/lib/auth.js'
+
+createCustomToken(principle_id: String, profile: Object)
+```
+
+## Implementation Details
+
+### Changes on SDK
+
+- Create custom token function for user server side use in py-skygear and js sdk, see sample code
+- Login with custom token for client side use in js, ios and android core sdk
+
+### Changes on API at skygear-server
+
+- New environment variable `CUSTOM_TOKEN_SECRET`
+    - Value will be used in user's server and skygear server, user's server create the token and skygear server verify it
+    - Need to be auto generated when create app in cloud
+- New handler `sso:custom_token:login` for sdk call
+    - Create skygear server token and login user
+    - Create new user when not found
+    - Will read the `skygear_profile` in jwt token, create profile for signup, update profile for login
+
+### Custom token claims (JWT)
+
+| Claims                                 | Description |
+|----------------------------------------|-------------|
+| sub (Subject)                          | Unique identifier of user in user's server, string between 1-36 characters long (Required) |
```

Comment:
Would we be more specific on what's UID? Is it the Skygear user ID? (But it is optional I suppose?)

Would we have another claims for the email address of the users?

Reply:
This should be user's auth server user id. I think user will be more convenient that they don't have to store extra id for this feature and reuse their user id. So we have not restricted the id to be uuid or skygear user id .

Reply:
The `skygear_profile` claim is for populating developer system user info to skygear. So the email address should be there. I will update the example